packages/dds/tree/src/feature-libraries/modular-schema/schemaChangeFamily.ts

```diff
@@ -0,0 +1,278 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { ChangeEncoder, ChangeFamily, JsonCompatibleReadOnly, ProgressiveEditBuilder } from "../../change-family";
+import { ChangeRebaser } from "../../rebase";
+import {
+    FieldSchema,
+    GlobalFieldKey,
+    LocalFieldKey,
+    SchemaData,
+    TreeSchema,
+    TreeSchemaIdentifier,
+} from "../../schema-stored";
+import { AnchorSet, Delta, FieldKey } from "../../tree";
+import { brand, Invariant, Multimap } from "../../util";
+import { FieldChangeHandler, FieldChangeMap, FieldChangeset, NodeChangeset } from "./fieldChangeHandler";
+import { FullSchemaPolicy } from "./fieldKind";
+
+export class SchemaChangeFamily implements
+    ChangeFamily<any, NodeChangeset>,
+    ChangeRebaser<NodeChangeset>,
+    ChangeEncoder<NodeChangeset> {
+    _typeCheck?: Invariant<any> | undefined;
+
+    constructor(
+        readonly rootSchema: TreeSchemaIdentifier | undefined,
+        readonly schemaData: SchemaData,
+        readonly schemaPolicy: FullSchemaPolicy,
+    ) { }
+
+    get rebaser(): ChangeRebaser<any> { return this; }
+    get encoder(): ChangeEncoder<any> { return this; }
+
+    compose(changes: NodeChangeset[]): NodeChangeset {
+        return this.composeI(this.rootSchema, changes);
+    }
+
+    private composeI(staticSchemaId: TreeSchemaIdentifier | undefined, changes: NodeChangeset[]): NodeChangeset {
+        if (changes.length === 1) {
+            return changes[0];
+        }
+
+        let schemaId = staticSchemaId;
+        const fieldChanges = new Multimap<FieldKey, FieldChangeset>();
+        for (const change of changes) {
+            if (change.schema !== undefined) {
+                schemaId = change.schema;
+            }
+
+            for (const key of Object.keys(change.fields)) {
+                fieldChanges.add(brand(key), change.fields[key]);
+            }
+        }
+
+        assert(schemaId !== undefined, "Need schema ID for root");
+        const schema = this.getTreeSchema(schemaId);
+
+        const composedFields: FieldChangeMap = {};
+        for (const field of fieldChanges.keys()) {
+            const childSchema = this.getChildSchema(schema, field);
+            const composedField = this.getChangeHandler(schema, field).rebaser.compose(
+                fieldChanges.get(field),
+                (...childChanges) => this.composeI(childSchema, childChanges),
+            );
+
+            // TODO: Could optimize by checking that composedField is non-empty
```

Comment:
We may want to make that check for the sake of making the code easier to test (so that test cases don't have to worry about empty lists).